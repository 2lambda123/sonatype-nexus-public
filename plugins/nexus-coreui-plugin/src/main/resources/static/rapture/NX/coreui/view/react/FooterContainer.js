/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Open Source Version is distributed with Sencha Ext JS pursuant to a FLOSS Exception agreed upon
 * between Sonatype, Inc. and Sencha Inc. Sencha Ext JS is licensed under GPL v3 and cannot be redistributed as part of a
 * closed source work.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
/*global Ext, NX, REACT_COMPONENTS, react, ReactDOM */

/**
 * React Footer Container
 *
 * @since 3.21
 */
Ext.define('NX.coreui.view.react.FooterContainer', {
  extend: 'Ext.Component',
  requires: [
    'NX.State',
    'NX.Permissions'
  ],

  alias: 'widget.nx-coreui-react-footer-container',

  reactView: undefined,

  reactViewProps: null,

  id: 'nxrm-react-footer-container',
  cls: 'nxrm-react-footer-container',

  listeners: {
    afterrender: 'initReactView',
    beforedestroy: 'destroyReactView'
  },

  initReactView: function() {
    const reactElement = react.createElement(this.reactView, {
      onClose: function() {
        this.dismissAlert()
      }.bind(this)
    }, null);
    ReactDOM.render(reactElement, this.getEl().dom);
  },

  destroyReactView: function() {
    if (this.reactView) {
      ReactDOM.unmountComponentAtNode(this.getEl().dom);
    }
  },

  refresh: function() {
    this.destroyReactView();
    this.initReactView();
    this.maybeSetVisible();
  },

  maybeSetVisible: function() {

    const isCircuitBreakerEnabled = NX.State.getValue('nexus.circuitb.enabled');
    const contentUsageEvaluationResults = NX.State.getValue('contentUsageEvaluationResult');
    const isHa = NX.State.getValue('nexus.datastore.clustered.enabled');
    const isProEdition = NX.State.getEdition() === 'PRO';
    const canViewMetrics = NX.Permissions.check('nexus:metrics:read');
    const softLimitCards = contentUsageEvaluationResults.filter(function(m) {
      return m.limitLevel === 'SOFT_LIMIT';
    });

    this.setVisible(!isProEdition && isCircuitBreakerEnabled && !isHa && softLimitCards.length > 0 && canViewMetrics);
    this.updateLayout();
  },

  dismissAlert: function(){
    this.getEl().child('div.nx-alert--warning').setDisplayed(false);
    this.updateLayout();
  },
});
