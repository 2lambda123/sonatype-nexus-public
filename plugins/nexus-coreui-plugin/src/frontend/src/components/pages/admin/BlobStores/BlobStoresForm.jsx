/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {useMachine} from '@xstate/react';
import {path} from 'ramda';
import {faTrash} from "@fortawesome/free-solid-svg-icons";

import {
  DynamicFormField,
  FieldWrapper,
  NxButton,
  NxCheckbox,
  NxErrorAlert,
  NxFontAwesomeIcon,
  NxTooltip,
  NxInfoAlert,
  Page,
  PageActions,
  PageHeader,
  PageTitle,
  Section,
  SectionFooter,
  Select,
  Textfield,
  Utils,
  NxForm,
  FormUtils
} from '@sonatype/nexus-ui-plugin';
import BlobStoresFormMachine from './BlobStoresFormMachine';
import UIStrings from '../../../../constants/UIStrings';
import CustomBlobStoreSettings from './CustomBlobStoreSettings';
import BlobStoreWarning from './BlobStoreWarning';

const FORM = UIStrings.BLOB_STORES.FORM;

export default function BlobStoresForm({itemId, onDone}) {
  const idParts = itemId.split('/');
  const pristineData = idParts?.length ? {
    type: idParts[0],
    name: idParts[1]
  } : {};

  const [current, send] = useMachine(BlobStoresFormMachine, {
    context: {
      pristineData
    },

    actions: {
      onCancel: onDone,
      onSaveSuccess: onDone,
      onDeleteSuccess: onDone
    },
    devTools: true
  });

  const isLoading = current.matches('loading');
  const isSaving = current.matches('saving');
  const isCreate = itemId === '';
  const isEdit = !isCreate;
  const {
    blobStoreUsage,
    data,
    isPristine,
    saveError,
    loadError,
    quotaTypes,
    repositoryUsage,
    type,
    types,
    validationErrors
  } = current.context;
  const isInvalid = Utils.isInvalid(validationErrors);
  const hasSoftQuota = path(['softQuota', 'enabled'], data);
  const cannotSave = isPristine || isInvalid;
  const cannotDelete = blobStoreUsage > 0 || repositoryUsage > 0;
  const deleteTooltip = cannotDelete ?
      UIStrings.BLOB_STORES.MESSAGES.CANNOT_DELETE(repositoryUsage, blobStoreUsage) :
      null;

  function setType(event) {
    send({type: 'SET_TYPE', value: event.currentTarget.value});
  }

  function updateField(event) {
    send({type: 'UPDATE', data: {[event.target.name]: event.target.value}});
  }

  function updateDynamicField(name, value) {
    send({
      type: 'UPDATE',
      data: {
        [name]: value}
    });
  }

  function toggleSoftQuota() {
    send({type: 'UPDATE_SOFT_QUOTA', name: 'enabled', value: !path(['softQuota', 'enabled'], data)});
  }

  function toggleCustomSetting(event) {
    send({type: 'UPDATE_CUSTOM_SETTINGS', name: event.target.id, value: null});
  }

  function updateCustomSettings(event) {
    const name = event.target.name;
    const value = event.target.value;
    send({type: 'UPDATE_CUSTOM_SETTINGS', name, value});
  }

  function updateQuotaField(event) {
    const name = event.target.name.replace('softQuota.', '');
    const value = event.target.value;
    send({type: 'UPDATE_SOFT_QUOTA', name, value});
  }

  function retry() {
    send({type: 'RETRY'});
  }

  function handleEnter(event) {
    if (event.key === 'Enter') {
      save(event);
    }
  }

  function save() {
    send({type: 'SAVE'});
  }

  function cancel() {
    onDone();
  }

  function confirmDelete() {
    send({type: 'CONFIRM_DELETE'});
  }

  function promoteToGroup() {
    send({type: 'PROMOTE_TO_GROUP'});
  }

  return <Page className="nxrm-blob-stores">
    {saveError && <NxErrorAlert>{saveError}</NxErrorAlert>}
    <PageHeader>
      <PageTitle text={isEdit ? FORM.EDIT_TILE(data.name) : FORM.CREATE_TITLE}
                 description={isEdit ? FORM.EDIT_DESCRIPTION(type?.id) : null}/>
      {isEdit && type?.id !== "Group" &&
      <PageActions>
        <NxButton variant="primary" onClick={promoteToGroup}>{FORM.PROMOTE_BUTTON}</NxButton>
      </PageActions>
      }
    </PageHeader>
    <NxForm className="nxrm-blob-stores-form"
            loading={isLoading}
            loadError={loadError ? `${loadError}` : null}
            doLoad={retry}
            onCancel={cancel}
            onSubmit={save}
            submitError={saveError}
            submitMaskState={isSaving ? false : null}
            submitMaskMessage={UIStrings.SAVING}
            submitBtnText={UIStrings.SETTINGS.SAVE_BUTTON_LABEL}
            validationErrors={FormUtils.saveTooltip({isPristine, isInvalid})}>
      <Section onKeyPress={handleEnter}>
        {isEdit && <NxInfoAlert>{FORM.EDIT_WARNING}</NxInfoAlert>}
        {isCreate &&
        <FieldWrapper labelText={FORM.TYPE.label} description={FORM.TYPE.sublabel}>
          <Select name="type" value={type?.id} onChange={setType}>
            <option value=""></option>
            {types.map(({id, name}) => <option key={id} value={id}>{name}</option>)}
          </Select>
        </FieldWrapper>
        }
        {type &&
        <>
          <BlobStoreWarning type={type}/>
          {isCreate &&
          <FieldWrapper labelText={FORM.NAME.label}>
            <Textfield {...Utils.fieldProps('name', current)} onChange={updateField}/>
          </FieldWrapper>
          }
          <CustomBlobStoreSettings type={type} current={current} updateCustomSettings={updateCustomSettings} toggleCustomSetting={toggleCustomSetting}/>
          {type?.fields?.map(field =>
              <FieldWrapper key={field.id}
                            labelText={field.label}
                            descriptionText={field.helpText}
                            isOptional={!field.required}>
                <DynamicFormField
                    id={field.id}
                    current={current}
                    initialValue={field.initialValue}
                    onChange={updateDynamicField}
                    dynamicProps={field}/>
              </FieldWrapper>
          )}
          <FieldWrapper labelText={FORM.SOFT_QUOTA.ENABLED.label} descriptionText={FORM.SOFT_QUOTA.ENABLED.sublabel}>
            <NxCheckbox {...Utils.checkboxProps(['softQuota', 'enabled'], current)} onChange={toggleSoftQuota}>
              {FORM.SOFT_QUOTA.ENABLED.text}
            </NxCheckbox>
          </FieldWrapper>
          {hasSoftQuota &&
          <>
            <FieldWrapper labelText={FORM.SOFT_QUOTA.TYPE.label} descriptionText={FORM.SOFT_QUOTA.TYPE.sublabel}>
              <Select {...Utils.fieldProps(['softQuota', 'type'], current)} onChange={updateQuotaField}>
                <option value="" disabled></option>
                {quotaTypes.map(({id, name}) => <option key={id} value={id}>{name}</option>)}
              </Select>
            </FieldWrapper>
            <FieldWrapper labelText={FORM.SOFT_QUOTA.LIMIT.label}>
              <Textfield {...Utils.fieldProps(['softQuota', 'limit'], current)} onChange={updateQuotaField}/>
            </FieldWrapper>
          </>
          }
          <SectionFooter>
            {itemId &&
            <NxTooltip title={deleteTooltip}>
              <NxButton variant="tertiary" className={cannotDelete && 'disabled'} onClick={confirmDelete}>
                <NxFontAwesomeIcon icon={faTrash}/>
                <span>{UIStrings.SETTINGS.DELETE_BUTTON_LABEL}</span>
              </NxButton>
            </NxTooltip>
            }
          </SectionFooter>
        </>
        }
      </Section>
    </NxForm>
  </Page>;
}
