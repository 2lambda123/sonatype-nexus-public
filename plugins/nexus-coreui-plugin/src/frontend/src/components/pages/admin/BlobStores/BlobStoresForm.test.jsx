/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import axios from 'axios';
import {when} from 'jest-when';
import {fireEvent, waitFor, waitForElementToBeRemoved} from '@testing-library/react'
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import TestUtils from '@sonatype/nexus-ui-plugin/src/frontend/src/interface/TestUtils';

import BlobStoresForm from './BlobStoresForm';

jest.mock('axios', () => ({
  ...jest.requireActual('axios'), // Use most functions from actual axios
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn()
}));

const blobstoreTypes = {
  data: [{
    "id": "File",
    "name": "File",
    "fields": [
      {
        "helpText": "An absolute path or a path relative to <data-directory>/blobs",
        "id": "path",
        "regexValidation": null,
        "required": true,
        "disabled": false,
        "readOnly": false,
        "label": "Path",
        "attributes": {
          tokenReplacement: "/<data-directory>/blobs/${name}",
          "long": true
        },
        "type": "string",
        "allowAutocomplete": false
      }
    ]
  }, {
    "id": "Group",
    "name": "Group",
    "fields": [
      {
        "id": "members",
        "required": true,
        "label": "Members",
        "initialValue": null,
        "attributes": {
          "toTitle": "Selected",
          "fromTitle": "Available",
          "buttons": ["up", "add", "remove", "down"],
          "options": ["default", "test", "test-promoted"]
        },
        "type": "itemselect",
        "allowAutocomplete": false
      }, {
        "id": "fillPolicy",
        "required": true,
        "label": "Fill Policy",
        "initialValue": null,
        "attributes": {
          "options": ["Round Robin", "Write to First"]
        },
        "type": "combobox"
      }
    ]
  }]
};

const quotaTypes = {
  data: [
    {
      "id": "spaceRemainingQuota",
      "name": "Space Remaining"
    }, {
      "id": "spaceUsedQuota",
      "name": "Space Used"
    }
  ]
};

describe('BlobStoresForm', function() {
  const onDone = jest.fn();

  function render(itemId) {
    return TestUtils.render(<BlobStoresForm itemId={itemId || ''} onDone={onDone}/>,
        ({getByRole, getByLabelText, queryByLabelText, getByText, queryByText}) => ({
          title: () => getByRole('heading', {level: 1}),
          typeSelect: () => queryByLabelText('Type'),
          name: () => queryByLabelText('Name'),
          path: () => getByLabelText('Path'),
          availableMembers: () => queryByLabelText('Available'),
          selectedMembers: () => queryByLabelText('Selected'),
          softQuota: () => getByLabelText('Soft Quota'),
          softQuotaType: () => queryByLabelText('Constraint Type'),
          softQuotaLimit: () => queryByLabelText('Constraint Limit (in MB)'),
          saveButton: () => getByText('Save'),
          cancelButton: () => getByText('Cancel'),
          promoteToGroup: () => queryByText('Promote to Group')
        }));
  }

  it('renders the loading spinner', async function() {
    axios.get.mockReturnValue(new Promise(() => {}));

    const {loadingMask} = render();

    expect(loadingMask()).toBeInTheDocument();
  });

  it('renders the type selection for create', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {loadingMask, typeSelect} = render();

    await waitForElementToBeRemoved(loadingMask);

    expect(typeSelect().options.length).toBe(3);
    expect(Array.from(typeSelect().options).map(option => option.textContent)).toEqual(expect.arrayContaining([
        '',
        'File',
        'Group'
    ]));
    expect(typeSelect()).toHaveValue('');
  });

  it ('renders the form and buttons when the File type is selected', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {container, loadingMask, typeSelect} = render();

    await waitForElementToBeRemoved(loadingMask);

    userEvent.selectOptions(typeSelect(), 'File');
    expect(typeSelect()).toHaveValue('File');

    expect(container).toMatchSnapshot();
  });

  it ('renders the name field and dynamic path field when the File type is selected', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {loadingMask, typeSelect, name, path} = render();

    await waitForElementToBeRemoved(loadingMask);

    userEvent.selectOptions(typeSelect(), 'File');
    expect(typeSelect()).toHaveValue('File');

    expect(name()).toBeInTheDocument();

    expect(path()).toBeInTheDocument();
    expect(path()).toHaveValue('/<data-directory>/blobs/');
  });

  it ('renders the soft quota fields when the blobstore type is selected', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {loadingMask, typeSelect, softQuota, softQuotaType, softQuotaLimit, getByText} = render();

    await waitForElementToBeRemoved(loadingMask);

    userEvent.selectOptions(typeSelect(), 'File');
    expect(typeSelect()).toHaveValue('File');

    expect(softQuota()).toBeInTheDocument();
    expect(softQuotaType()).not.toBeInTheDocument();
    expect(softQuotaLimit()).not.toBeInTheDocument();

    fireEvent.click(softQuota());

    expect(softQuotaType()).toBeInTheDocument();
    expect(softQuotaLimit()).toBeInTheDocument();
  });

  it('enables the save button when there are changes', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {loadingMask, typeSelect, name, softQuota, softQuotaType, softQuotaLimit, saveButton} = render();

    await waitForElementToBeRemoved(loadingMask);

    userEvent.selectOptions(typeSelect(), 'File');
    expect(typeSelect()).toHaveValue('File');

    expect(saveButton()).toHaveClass('disabled');

    userEvent.type(name(), 'test');
    expect(name()).toHaveValue('test');

    expect(saveButton()).not.toHaveClass('disabled');

    fireEvent.click(softQuota());

    expect(saveButton()).toHaveClass('disabled');

    userEvent.selectOptions(softQuotaType(), 'spaceRemainingQuota');
    expect(softQuotaType()).toHaveValue('spaceRemainingQuota');
    userEvent.type(softQuotaLimit(), '100');
    expect(softQuotaLimit()).toHaveValue('100');

    expect(saveButton()).not.toHaveClass('disabled');

    userEvent.clear(softQuotaLimit());
    expect(softQuotaLimit()).toHaveValue('');

    expect(saveButton()).toHaveClass('disabled');

    fireEvent.click(softQuota());

    expect(saveButton()).not.toHaveClass('disabled');
  });

  it('creates a new file blob store', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);

    const {loadingMask, typeSelect, name, path, softQuota, softQuotaType, softQuotaLimit, saveButton} = render();

    await waitForElementToBeRemoved(loadingMask);

    userEvent.selectOptions(typeSelect(), 'File');
    expect(typeSelect()).toHaveValue('File');
    userEvent.type(name(), 'test');
    expect(name()).toHaveValue('test');
    expect(path()).toHaveValue('/<data-directory>/blobs/test');
    userEvent.clear(path());
    userEvent.type(path(), 'testPath');
    expect(path()).toHaveValue('testPath');
    userEvent.click(softQuota());
    userEvent.selectOptions(softQuotaType(), 'spaceRemainingQuota');
    expect(softQuotaType()).toHaveValue('spaceRemainingQuota');
    userEvent.type(softQuotaLimit(), '100');
    expect(softQuotaLimit()).toHaveValue('100');
    userEvent.click(saveButton());

    expect(axios.post).toHaveBeenCalledWith(
        '/service/rest/v1/blobstores/file',
        {
          name: 'test',
          path: 'testPath',
          softQuota: {
            enabled: true,
            type: 'spaceRemainingQuota',
            limit: '100'
          }
        }
    );
  });

  it('edits a file blob store', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);
    when(axios.get).calledWith('/service/rest/v1/blobstores/file/test').mockResolvedValue({
      data: {
        path: 'testPath',
        softQuota: {
          type: 'spaceRemainingQuota',
          limit: '100'
        }
      }
    });

    const {
      loadingMask,
      promoteToGroup
    } = render('file/test');

    await waitForElementToBeRemoved(loadingMask);

    expect(promoteToGroup()).toBeInTheDocument();
  });

  it('edits a file blob store', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);
    when(axios.get).calledWith('/service/rest/v1/blobstores/file/test').mockResolvedValue({
      data: {
        path: 'testPath',
        softQuota: {
          type: 'spaceRemainingQuota',
          limit: '100'
        }
      }
    });

    const {
      getByText,
      loadingMask,
      name,
      path,
      title,
      softQuota,
      softQuotaType,
      softQuotaLimit,
      typeSelect
    } = render('file/test');

    await waitForElementToBeRemoved(loadingMask);

    expect(title()).toHaveTextContent('Edit test');
    expect(getByText('File Blob Store')).toBeInTheDocument();

    // The type and name fields cannot be changed during edit
    expect(typeSelect()).not.toBeInTheDocument();
    expect(name()).not.toBeInTheDocument();

    expect(path()).toHaveValue('testPath');
    expect(softQuota()).toBeChecked();
    expect(softQuotaType()).toHaveValue('spaceRemainingQuota');
    expect(softQuotaLimit()).toHaveValue('100');
  });

  it('edits a group blob store', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);
    when(axios.get).calledWith('/service/rest/v1/blobstores/group/test').mockResolvedValue({
      data: {
        "softQuota" : null,
        "members" : [ "test-promoted" ],
        "fillPolicy" : "writeToFirst"
      }
    });

    const {
      availableMembers,
      getByText,
      loadingMask,
      name,
      selectedMembers,
      title,
      typeSelect
    } = render('group/test');

    await waitForElementToBeRemoved(loadingMask);

    expect(title()).toHaveTextContent('Edit test');
    expect(getByText('Group Blob Store')).toBeInTheDocument();

    expect(typeSelect()).not.toBeInTheDocument();
    expect(name()).not.toBeInTheDocument();

    expect(availableMembers()).toContainElement(getByText('default'));
    expect(selectedMembers()).toContainElement(getByText('test-promoted'));
  });

  it('promote to group is not shown when editing a group', async function() {
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/types').mockResolvedValue(blobstoreTypes);
    when(axios.get).calledWith('/service/rest/internal/ui/blobstores/quotaTypes').mockResolvedValue(quotaTypes);
    when(axios.get).calledWith('/service/rest/v1/blobstores/group/test').mockResolvedValue({
      data: {
        "softQuota" : null,
        "members" : [ "test-promoted" ],
        "fillPolicy" : "writeToFirst"
      }
    });

    const {
      loadingMask,
      promoteToGroup
    } = render('group/test');

    await waitForElementToBeRemoved(loadingMask);

    expect(promoteToGroup()).not.toBeInTheDocument();
  });
});
