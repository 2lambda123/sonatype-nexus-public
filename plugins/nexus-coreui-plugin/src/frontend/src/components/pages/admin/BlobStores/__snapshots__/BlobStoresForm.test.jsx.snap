// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlobStoresForm renders the form and buttons when the File type is selected 1`] = `
<div>
  <main
    class="nx-page-main nxrm-blob-stores"
  >
    <div
      class="nxrm-page-header"
    >
      <div
        class="nx-page-title"
      >
        <h1
          class="nx-h1 nx-feature-name"
        >
          <span>
            Create Blob Store
          </span>
        </h1>
      </div>
    </div>
    <div
      class="nxrm-content-body nxrm-blob-stores-form"
    >
      <div
        class="nxrm-section nx-tile"
      >
        <div
          class="nx-tile-content"
        >
          <div
            class="nx-form-group"
          >
            <label
              class="nx-label"
              for="type"
            >
              <span
                class="nx-label__text"
              >
                Type
              </span>
            </label>
            <div
              class="nxrm-select"
            >
              <select
                id="type"
                name="type"
              >
                <option
                  value=""
                />
                <option
                  value="File"
                >
                  File
                </option>
              </select>
            </div>
          </div>
          <div
            class="nx-form-group"
          >
            <label
              class="nx-label"
              for="name"
            >
              <span
                class="nx-label__text"
              >
                Name
              </span>
            </label>
            <div
              class="nx-text-input pristine"
            >
              <div
                class="nx-text-input__box"
              >
                <input
                  aria-errormessage="nx-text-input-invalid-message-00000000"
                  aria-invalid="false"
                  class="nx-text-input__input"
                  id="name"
                  name="name"
                  type="text"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-check fa-w-16 nx-icon nx-icon--valid nx-icon"
                  data-icon="check"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon nx-icon--invalid nx-icon"
                  data-icon="exclamation-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <div
                class="nx-text-input__invalid-message"
                id="nx-text-input-invalid-message-00000000"
                role="alert"
              />
            </div>
          </div>
          <div
            class="nx-form-group"
          >
            <label
              class="nx-label"
              for="path"
            >
              <span
                class="nx-label__text"
              >
                Path
              </span>
            </label>
            <p
              class="nx-p"
            >
              An absolute path or a path relative to &lt;data-directory&gt;/blobs
            </p>
            <div
              class="nx-text-input nx-text-input--long pristine"
            >
              <div
                class="nx-text-input__box"
              >
                <input
                  aria-errormessage="nx-text-input-invalid-message-11111111"
                  aria-invalid="false"
                  class="nx-text-input__input"
                  id="path"
                  name="path"
                  type="text"
                  value="/<data-directory>/blobs/"
                />
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-check fa-w-16 nx-icon nx-icon--valid nx-icon"
                  data-icon="check"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                    fill="currentColor"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon nx-icon--invalid nx-icon"
                  data-icon="exclamation-circle"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <div
                class="nx-text-input__invalid-message"
                id="nx-text-input-invalid-message-11111111"
                role="alert"
              />
            </div>
          </div>
          <div
            class="nx-form-group"
          >
            <label
              class="nx-label"
              for="softQuota.enabled"
            >
              <span
                class="nx-label__text"
              >
                Soft Quota
              </span>
            </label>
            <p
              class="nx-p"
            >
              Raises an alert when the blob store exceeds a constraint
            </p>
            <label
              class="nx-radio-checkbox nx-checkbox tm-unchecked"
              name="softQuota.enabled"
            >
              <input
                class="nx-radio-checkbox__input nx-checkbox__input"
                id="softQuota.enabled"
                type="checkbox"
              />
              <span
                class="nx-radio-checkbox__control nx-checkbox__box"
              >
                Â 
              </span>
              <span
                aria-label=""
                class="nx-radio-checkbox__content nx-checkbox__content"
                title=""
              >
                Enabled
              </span>
            </label>
          </div>
          <div
            class="nx-footer"
          >
            <button
              aria-disabled="true"
              aria-label="There are no changes"
              class="nx-btn disabled nx-btn--primary"
              title="There are no changes"
              type="submit"
            >
              Save
            </button>
            <button
              class="nx-btn"
            >
              Cancel
            </button>
            
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
